### ISSUE TITLE
Requirements check failed for JDK 8 ('1.8.*')! Detected version: null

### ISSUE THREAD

**Issue by maniveltvl** [apache/cordova-android#1040]
### Issue Type
Requirements check failed for JDK 8 ('1.8.*')! Detected version: null

- [x] Bug Report
- [x] Feature Request
- [ ] Support Question

I want to upgrade my Android SDK version API Level 28 to 29. So I have added Android platform  version 9.0.0. Then I try to build the APK but  I am getting the below error.
'Requirements check failed for JDK 8 ('1.8.*')! Detected version: null
Check your ANDROID_SDK_ROOT / JAVA_HOME / PATH environment variables.'

## Information
1. cordova create hello com.example.hello HelloWorld
2. cd hello
3. cordova platform add android@9.0.0
4. cordova build android

### Command or Code
I found that check_req.js files line no 223 returns output as null. 

### Environment, Platform, Device
Platform: Android.

### Version information
What are relevant versions you are using?
For example:
Cordova: 8.1.0

**Comment by breautek** [comment]
This suggests you don't have a java JDK installed. If you do, you should ensure the `JAVA_HOME` environment variable is pointing to the proper location.

**Comment by paulodetarsofm** [comment]
I'm facing the same problem. I have everything configured, the weird is I started to get this issue only after restart my computer, during the installation of Android Studio/SDK and creating the Cordova app for the first time, worked fine. I'm looking for a solution, no success. I reinstalled everything: Android Studio, JDK, Cordova, but nothing...

I have all these environment variables defined, but the Java version can't be detected by Cordova. If I run java -version, I get:

java version "1.8.0_261"
Java(TM) SE Runtime Environment (build 1.8.0_261-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.261-b12, mixed mode)

So, I have the Java installed! What is happening? What is the problem?

**Comment by breautek** [comment]
For linux/mac try running:

`echo $JAVA_HOME`

For windows i think it's:

`echo %ANDROID_HOME%`

But I'm not familiar with windows shell programming, so I may be recalling incorrectly.

If the shell prints out empty  string/line, then you should set the `JAVA_HOME` environment variable.

Learn how via: https://cordova.apache.org/docs/en/latest/guide/platforms/android/index.html#setting-environment-variables

**Comment by paulodetarsofm** [comment]
I'm using Windows 10. The variables are defined correctly:

System Variables:
![image](https://user-images.githubusercontent.com/4804498/89743455-432c8a80-da9b-11ea-957c-e0928557cc60.png)

Path variable:
![image](https://user-images.githubusercontent.com/4804498/89743433-11b3bf00-da9b-11ea-9151-da451202adb9.png)

**Comment by maniveltvl** [comment]
Hi breautek,

I have installed JDK and i have set JAVA_HOME variable in environment variable but still it is not working. Then I have found that some issue in  check_req.js.

I have changed code from 
`return execa('javac', ["-version"], { all: true })`
to 
`return execa('javac -version', [], { all: true })` in line no 223

After changed to the above code able to get java version and working fine.

Can you please try to reproduce the issue and change to the above working code.

**Comment by breautek** [comment]
Master currently has https://github.com/apache/cordova-android/blob/master/bin/templates/cordova/lib/check_reqs.js#L211 and so does `v9.0.0` release: https://github.com/apache/cordova-android/blob/rel/9.0.0/bin/templates/cordova/lib/check_reqs.js

Are you sure you're up to date?

Try using `cordova@10`, instead of `cordova@8.1.0`

**Comment by maniveltvl** [comment]
Hi @breautek 

Please find my Cordova version below
`cordova -v
8.1.0`

But I want to install android@9.0.0 to get Android API level 29. When I install android@9.0.0, i am getting the above error. So as of now i am changing the below code to work

`return execa('javac -version', [], { all: true })'

**Comment by breautek** [comment]
> execa('javac -version', [], { all: true })

This should only work if executing in shell mode. It doesn't follow the documentation of [execa](https://www.npmjs.com/package/execa)

In general we refrain from using shell modes because it's slower, not cross-platform, and easy to introduce security vulnerabilities if care is not taken.

Additionally I haven't been able to reproduce this using the latest versions of the software.

The latest versions of cordova to date are:
`cordova@10`
`cordova-android@9`

cordova 8.1.0 is an old and no longer maintained version. If you can reproduce your issue on cordova 10 / cordova-android@9, then do let me know.

**Comment by MatHackDev** [comment]
Hi everyone;

I had the same problem on my Mac; after analyzing my package-lock.json; I had a version of execa:1.0.0 ! I don't know where it comes from.

I removed my package-lock.json and the node_modules, then I did "npm i -D execa". Which will now be in 4.0.3.

That solved my problem.

**Comment by erisu** [comment]
@MatHackDev I think all you needed to do was remove the `package-lock.json` & `node_modules` directory and run `npm i` to re-checkout the new package. I don't think you need to run `npm i -D execa` because this would install it at the project level and most likely not be used by `cordova-android` build scripts.

**Comment by MatHackDev** [comment]
Thanks for answer @erisu but "remove the package-lock.json & node_modules directory and run npm i" it's not working for me. But i will test "remove platforms and run cordova prepare."

Maybe it's related to another package that creates a conflict.

**Comment by ajberasategui** [comment]
I experienced the same issue. I'm using CLI v9. Even if it does not make sense to me at all, deleting package-lock and node_modules and reinstalling did the trick. 😕

**Comment by orkuncaylar** [comment]
> Hi everyone;
> 
> I had the same problem on my Mac; after analyzing my package-lock.json; I had a version of execa:1.0.0 ! I don't know where it comes from.
> 
> I removed my package-lock.json and the node_modules, then I did "npm i -D execa". Which will now be in 4.0.3.
> 
> That solved my problem.
thanks!!!!!

**Comment by breautek** [comment]
I think it's time to close this issue. The problem seems to be in the `execa` package, which is a sub-dependency and the problem is already fixed, but depending on when you installed Cordova, you may have an old version of `execa` causing this problem. As many others stated, removing the `node_modules` and `package-lock.json` file to force update all modules on `npm install` .

**Comment by thibaudsowa** [comment]
Hi, here is my contribution.

I have exactly the same problem, after hours of research I think I have more information about this.
Actually, I discovered that the problem came from `execa` installed by other dependencies, there is a conflict between different versions.

Here is my dependency tree for `execa`:
```
$ npm ls execa
myproject@0.0.1 C:\somepath
+-- @angular-devkit/build-angular@0.13.9
| `-- webpack-dev-server@3.1.14
|   +-- internal-ip@3.0.1
|   | `-- default-gateway@2.7.2
|   |   `-- execa@0.10.0
|   `-- yargs@12.0.2
|     `-- os-locale@3.1.0
|       `-- execa@1.0.0
+-- @angular/compiler-cli@7.2.15
| `-- yargs@9.0.1
|   `-- os-locale@2.1.0
|     `-- execa@0.7.0
+-- @ionic/cli@6.11.11
| `-- os-name@4.0.0
|   `-- windows-release@4.0.0
|     `-- execa@4.0.3
+-- cordova-android@9.0.0
| `-- execa@4.0.3
`-- protractor@5.4.4
  `-- yargs@12.0.5
    `-- os-locale@3.1.0
      `-- execa@1.0.0
```

When there is different version of a library, npm create sub `node_modules` directory:

The version of `execa` in `/node_modules/execa` is 0.10.0
The version of `execa` in `/node_modules/cordova-android/node_modules/execa` is 4.0.3

I don't know why but when I execute `cordova build android` the `execa` used is the first one (the bad one...)
I figured out by editing both lib file with some log.

Of course I did the classic "delete `node_modules` + `package-lock.json`" without no success.
The only workaround for me is to add `execa` in my own dependency, but I'm not happy with that because I don't use it and I'm afraid to delete it some day when I will do some dependency cleaning...)

Thanks for your help!

I'm on windows 10, cordova 10.0.0, cordova-android 9.0.0, npm 6.14.5, node 10.18.1

**Comment by breautek** [comment]
> I don't know why but when I execute cordova build android the execa used is the first one (the bad one...)
I figured out by editing both lib file with some log.

I think you should open a new issue and fill out the issue form documenting how `cordova-android` appears to be using the wrong `execa`. I also have a theory that you can add... 

That theory is because our executing code that depends on `execa` doesn't actually live inside `node_modules/cordova-android` and is instead copied into your project folder (`platforms/android/`). So this causes the NPM/Node module resolution to break. Consequentially, it also causes the problem to be fixed if you add it as your own dependency.

**Comment by premktr007** [comment]
try downgrading the android version. it worked for I also faced the same problem

**Comment by dahlseng** [comment]
I Deleted J 15 in library and started build and it started :)

**Comment by rkmaxwell** [comment]
Okay so this is how i solved it:  I am using jdk 1.8u 261, so i therefore **uninstalled** any other version below this version, close the CLI and reopen  then run -cordova build android   if you added android platform or cordova build ios for ios. 
![image](https://user-images.githubusercontent.com/52816026/112297794-58ab9900-8ca7-11eb-904e-62626013a0c8.png)

**Comment by fbl773** [comment]
I had an issue similar to this, my version check failed because I am running with JDK 15.0. 
I saw where the error was coming from so I modified the file at `platforms/android/cordova/lib/check_reqs.js` to have 
`EXPECTED_JAVA_VERSION = "15.x"` rather than `1.8.x` and everything ran fine... Why is it still dependent on JDK 1.8? Should this not have worked?

**Comment by breautek** [comment]
> EXPECTED_JAVA_VERSION = "15.x" rather than 1.8.x and everything ran fine... Why is it still dependent on JDK 1.8? Should this not have worked?

Android SDK & Android Gradle Plugin only officially supports Java 7 and 8. Last time I tested on anything higher, you could build but not all of the android tooling works on Java > 8.

Android will be officially supporting Java 11 *soon* (not sure when it will be released but Java11 support is available in their prerelease)

**Comment by lihas** [comment]
In my case I had "C:\Program Files\Common Files\Oracle\Java\javapath" in system path before the path to the jdk I wanted cordova to see -C:\Program Files\Java\jdk1.8.0_291\bin
Switched the order and this error went away.

**Comment by iAbran** [comment]
> Master currently has https://github.com/apache/cordova-android/blob/master/bin/templates/cordova/lib/check_reqs.js#L211 and so does `v9.0.0` release: https://github.com/apache/cordova-android/blob/rel/9.0.0/bin/templates/cordova/lib/check_reqs.js
> 
> Are you sure you're up to date?
> 
> Try using `cordova@10`, the current latest of the cli, then reinstall the android platform:
> 
> ```
> cordova platform remove android
> cordova platform add android
> ```

Thanks so much, unistall and install work for me. Love you bro
